### PR TITLE
SCP-259 onboard artifact management to flink cli

### PR DESCRIPTION
**Pull request by byang-confluent**
Release Notes
-------------

Breaking Changes

New Features
[Early Access] 
- New Flink artifact lifecycle management command for Flink EA customers. 

Bug Fixes
- PLACEHOLDER

Checklist
---------
- [ ] Leave this box unchecked if features are not yet available in production

What
----

[Early Access]

This PR adds new Flink artifact lifecycle management command for Flink EA customers.  

All existing flink commands are already gated with a feature flag cli.flink ([LD flag](https://app.launchdarkly.com/cli/production/features/cli.flink/targeting)). This new flink artifact command will use the same flag for access control.

```bash
# create
confluent flink artifact create <artifact-name> --artifact-file <local-file-name>

# delete 
confluent flink artifact delete <plugin-id>

# Describe
confluent flink artifact describe <plugin-id>

# List
confluent flink artifact list
```

References
----------

Test & Review
-------------

### CODE REVIEW

File: internal/flink/command_artifact.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	Id             string `serialized:"id"`
+	ConnectorClass string `serialized:"connector_class"`
+	ContentFormat  string `serialized:"connector_class"`
+	Scope          string
+}
+
+type customPluginHumanOut struct {
+	Name           string `human:"Name"`
+	Id             string `human:"Plugin ID"`
+	ConnectorClass string `human:"Version ID"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:23]: Why is this called "ConnectorClass" but prints as "Version ID"? Are those the same thing?

File: internal/flink/command_artifact.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	pcmd "github.com/confluentinc/cli/v3/pkg/cmd"
+	"github.com/confluentinc/cli/v3/pkg/output"
+)
+
+type customPluginSerializedOut struct {
+	Name           string `serialized:"name"`
+	Id             string `serialized:"id"`
+	ConnectorClass string `serialized:"connector_class"`
+	ContentFormat  string `serialized:"connector_class"`
+	Scope          string
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:17]: Why isn't this serialized?

File: internal/flink/command_artifact.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+	pcmd "github.com/confluentinc/cli/v3/pkg/cmd"
+	"github.com/confluentinc/cli/v3/pkg/output"
+)
+
+type customPluginSerializedOut struct {
+	Name           string `serialized:"name"`
+	Id             string `serialized:"id"`
+	ConnectorClass string `serialized:"connector_class"`
+	ContentFormat  string `serialized:"connector_class"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:16]: This has the same serialized name

File: internal/flink/command_artifact.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	cmd := &cobra.Command{
+		Use:         "artifact",
+		Short:       "Manage Flink UDF artifact",
+		Annotations: map[string]string{pcmd.RunRequirement: pcmd.RequireNonAPIKeyCloudLogin},
+	}
+
+	cmd.AddCommand(c.newCreateCommand())
+	cmd.AddCommand(c.newDescribeCommand())
+	cmd.AddCommand(c.newDeleteCommand())
+	cmd.AddCommand(c.newListCommand())
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:38]:
```suggestion
	cmd.AddCommand(c.newListCommand())
	
```

File: internal/flink/command_artifact.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+func printTable(cmd *cobra.Command, plugin connectcustompluginv1.ConnectV1CustomConnectorPlugin) error {
+	table := output.NewTable(cmd)
+	if output.GetFormat(cmd) == output.Human {
+		table.Add(&customPluginHumanOut{
+			Id:             plugin.GetId(),
+			Name:           plugin.GetDisplayName(),
+			ConnectorClass: plugin.GetConnectorClass(),
+			ContentFormat:  plugin.GetContentFormat(),
+			Scope:          "org",
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:50]: Why doesn't the backend return this value?

File: internal/flink/command_artifact_create.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/confluentinc/cli/v3/pkg/examples"
+	"github.com/confluentinc/cli/v3/pkg/output"
+)
+
+type pluginCreateOut struct {
+	Name           string `human:"Name" serialized:"name"`
+	Id             string `human:"Plugin ID" serialized:"id"`
+	ConnectorClass string `human:"Version ID" serialized:"connector_class"`
+	ContentFormat  string `human:"Content Format" serialized:"content_format"`
+	Scope          string `human:"Scope"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:22]: Why don't we have a serialized value?

File: internal/flink/command_artifact_create.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		),
+	}
+
+	cmd.Flags().String("artifact-file", "", "ZIP/JAR Flink UDF artifact file.")
+	cmd.Flags().String("description", "", "Description of Flink UDF artifact.")
+	pcmd.AddContextFlag(cmd, c.CLICommand)
+	pcmd.AddOutputFlag(cmd)
+
+	cobra.CheckErr(cmd.MarkFlagRequired("artifact-file"))
+	cobra.CheckErr(cmd.MarkFlagFilename("artifact-file", "jar"))
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:46]: The flag description implies this can also be a `.zip` file

File: internal/flink/utils.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		return err
+	}
+
+	if err := writer.Close(); err != nil {
+		return err
+	}
+
+	client := &http.Client{
+		Timeout: 20 * time.Minute,
+	}
```
Comment (brianstrauch) [inline internal/flink/utils.go:45]: Wow. How big are these files?

Comment (brianstrauch) [reply internal/flink/utils.go:45]:
```suggestion
	client := &http.Client{Timeout: 20 * time.Minute}
```

File: internal/flink/utils.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	_, err = sling.New().Client(client).Base(url).Set("Content-Type", writer.FormDataContentType()).Post("").Body(&buffer).ReceiveSuccess(nil)
```
Comment (brianstrauch) [inline internal/flink/utils.go:46]: Why aren't we using the SDK here?

**Review by brianstrauch — CHANGES_REQUESTED** [review]

File: internal/flink/command_artifact.go (hunk 1 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go:23]: @brianstrauch , this is a good question. 

we will reuse existing custom connector plugin api for flink artifact. This is just for EA only. Post EA, flink team should propose finalized public api for flink artifact management. 

In order to reuse custom connector plugin api to fit flink EA scenario, we return `versionid` from `connectClass` field for flink-udf case. This will change when Flink GA'ed.

File: internal/flink/command_artifact.go (hunk 2 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go:17]: `scope` is not a part of response payload for custom connector api. but we will need this info for flink case. hense we populate it for EA only

File: internal/flink/command_artifact.go (hunk 3 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go:16]: good catch!

File: internal/flink/command_artifact.go (hunk 5 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go:50]: scope is not a part of response payload for custom connector api. but we will need this info for flink case. hense we populate it for EA only

File: internal/flink/command_artifact_create.go (hunk 7 above)
Comment (byang-confluent) [reply internal/flink/command_artifact_create.go:46]: good catch. updated.

File: internal/flink/utils.go (hunk 8 above)
Comment (byang-confluent) [reply internal/flink/utils.go:45]: normally it's <20MB in terms of file size. this time is cloned from https://github.com/confluentinc/cli/blob/3b0527b316868d961060f38b8a47e89f7ecd2a1e/internal/connect/utils.go#L118-L119

File: internal/flink/utils.go (hunk 9 above)
Comment (byang-confluent) [reply internal/flink/utils.go:46]: we are using aws presigned url for plugin artifact upload, which requires either aws sdk nor custom connector plugin sdk. this is also clone from https://github.com/confluentinc/cli/blob/3b0527b316868d961060f38b8a47e89f7ecd2a1e/internal/connect/utils.go#L120-L121

File: internal/flink/command_artifact.go (hunk 1 above)
Comment (pgaref) [reply internal/flink/command_artifact.go:23]: Lets at least add some TODOs for all these rough edges until then please

Comment (byang-confluent) [reply internal/flink/command_artifact.go:23]: @pgaref do you have workitem to create flink artifact api post EA? i can link them here otherwise i'll create one

Comment (pgaref) [reply internal/flink/command_artifact.go:23]: Lets use https://confluentinc.atlassian.net/browse/FRT-334

File: internal/flink/command_artifact_create.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	if err := uploadFile(resp.GetUploadUrl(), pluginFileName, resp.GetUploadFormData()); err != nil {
+		return err
+	}
+
+	createArtifactRequest := connectcustompluginv1.ConnectV1CustomConnectorPlugin{
+		DisplayName:   connectcustompluginv1.PtrString(displayName),
+		Description:   connectcustompluginv1.PtrString(description),
+		ConnectorType: connectcustompluginv1.PtrString("flink-udf"),
```
Comment (AlanConfluent) [inline internal/flink/command_artifact_create.go:82]: Is this the type that differentiates it as flink?

Comment (byang-confluent) [reply internal/flink/command_artifact_create.go:82]: yes, correct, we track type as "flink-udf" internally. for custom connector it is "connector"

File: internal/flink/command_artifact.go (hunk 2 above)
Comment (brianstrauch) [reply internal/flink/command_artifact.go:17]: > scope is not a part of response payload for custom connector api

Why not?

Comment (byang-confluent) [reply internal/flink/command_artifact.go]: flink udf is reusing exsisting custom connector plugin api for EA users.

File: internal/flink/command_artifact.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+// This is reuse existing custom connector plugin api for flink udf management for EA customer only
+//
+//	aka, `ConnectorClass` will return version ID for EA
+//	For flink GA, flink team will have public API to do so
+type flinkArtifactHumanOut struct {
+	Name           string `human:"Name"`
+	Id             string `human:"Plugin ID"`
+	ConnectorClass string `human:"Version ID"`
+	ContentFormat  string `human:"Content Format"`
+}
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:29]: These structs can be combined into one since all types are strings

File: internal/flink/command_artifact.go (hunk 1 above)
Comment (brianstrauch) [reply internal/flink/command_artifact.go:27]: Should the serialized output also say "version_id"?

File: internal/flink/command_artifact_create.go (hunk 7 above)
Comment (brianstrauch) [reply internal/flink/command_artifact_create.go:45]: Did you remember to commit this change?

File: internal/flink/command_artifact_create.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+		UploadSource: &connectcustompluginv1.ConnectV1CustomConnectorPluginUploadSourceOneOf{
+			ConnectV1UploadSourcePresignedUrl: connectcustompluginv1.NewConnectV1UploadSourcePresignedUrl("PRESIGNED_URL_LOCATION", resp.GetUploadId()),
+		},
+	}
+
+	pluginResp, err := c.V2Client.CreateCustomPlugin(createArtifactRequest)
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:87]:
```suggestion
	plugin, err := c.V2Client.CreateCustomPlugin(createArtifactRequest)
```

File: internal/flink/command_artifact_create.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	return cmd
+}
+
+func (c *command) createArtifact(cmd *cobra.Command, args []string) error {
+	displayName := args[0]
+	description, err := cmd.Flags().GetString("description")
+	if err != nil {
+		return err
+	}
+	pluginFileName, err := cmd.Flags().GetString("artifact-file")
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:56]:
```suggestion

	artifactFile, err := cmd.Flags().GetString("artifact-file")
```

File: internal/flink/command_artifact_create.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+		return err
+	}
+
+	extension := strings.ToLower(strings.TrimPrefix(filepath.Ext(pluginFileName), "."))
+	if extension != "jar" {
+		return fmt.Errorf(`only file extensions ".jar" is allowed`)
+	}
+
+	request := *connectcustompluginv1.NewConnectV1PresignedUrlRequest()
+	request.SetContentFormat(extension)
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:67]: Can you initialize this struct directly?

File: internal/flink/command_artifact_describe.go [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+func (c *command) newDescribeCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "describe <id>",
+		Short: "Describe a Flink UDF artifact.",
+		Args:  cobra.ExactArgs(1),
+		RunE:  c.describe,
+		Example: examples.BuildExampleString(
+			examples.Example{
+				Text: "Describe Flink UDF artifact",
+				Code: "confluent connect artifact describe ccp-123456",
```
Comment (brianstrauch) [inline internal/flink/command_artifact_describe.go:19]:
```suggestion
				Code: "confluent flink artifact describe ccp-123456",
```

Comment (brianstrauch) [reply internal/flink/command_artifact_describe.go:19]: Why does the resource start with "ccp-" if this is a Flink command?

Comment (byang-confluent) [reply internal/flink/command_artifact_describe.go:19]: as we reusing custom connector plugin api, plugin id will start with `ccp-` for EA

**Comment by byang-confluent** [comment]
| Did you remember to commit this change?
@brianstrauch  flink artifact will be JAR only for EA

File: internal/flink/command_artifact.go (hunk 11 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go]: i will keep them as they are as it looks more readable. wdyt?

File: internal/flink/command_artifact_create.go (hunk 7 above)
Comment (byang-confluent) [reply internal/flink/command_artifact_create.go:45]: @brianstrauch flink artifact will be JAR only for EA

Comment (byang-confluent) [reply internal/flink/command_artifact_create.go:45]: updated by removing `zip`

File: internal/flink/command_artifact.go (hunk 11 above)
Comment (brianstrauch) [reply internal/flink/command_artifact.go]: Let's combine them to make it easier to update them in the future and more closely align with our other commands

Comment (byang-confluent) [reply internal/flink/command_artifact.go]: updated

File: internal/flink/command_artifact.go [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	"github.com/spf13/cobra"
+
+	connectcustompluginv1 "github.com/confluentinc/ccloud-sdk-go-v2/connect-custom-plugin/v1"
+
+	pcmd "github.com/confluentinc/cli/v3/pkg/cmd"
+	"github.com/confluentinc/cli/v3/pkg/output"
+)
+
+// TODO: https://confluentinc.atlassian.net/browse/FRT-334
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:12]: This is a source-available repository, so can you summarize this ticket since external contributors aren't able to access it?

File: internal/flink/command_artifact.go [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+// This is reuse existing custom connector plugin api for flink udf management for EA customer only
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:13]:
```suggestion
// This is reuse existing custom connector plugin API for Flink UDF management for EA customer only
```

Comment (brianstrauch) [reply internal/flink/command_artifact.go:13]: Not sure what this comment means?

File: internal/flink/command_artifact.go [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+//
+//	aka, `ConnectorClass` will return version ID for EA
+//	For flink GA, flink team will have public API to do so
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:16]:
```suggestion
//	For Flink GA, Flink team will have public API to do so
```

File: internal/flink/command_artifact.go [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+type flinkArtifactOut struct {
+	Name           string `serialized:"name" human:"Name"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:18]: Can you put human before serialized like in our other commands?

File: internal/flink/command_artifact.go [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+	Id             string `serialized:"plugin_id" human:"Plugin ID"`
+	ConnectorClass string `serialized:"version_id" human:"Version ID"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:20]:
```suggestion
	VersionId      string `serialized:"version_id" human:"Version ID"`
```

File: internal/flink/command_artifact.go [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+	ContentFormat  string `serialized:"content_format" human:"Content Format"`
+}
+
+func (c *command) newArtifactCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:         "artifact",
+		Short:       "Manage Flink UDF artifact",
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:27]:
```suggestion
		Short:       "Manage Flink UDF artifacts.",
```

File: internal/flink/command_artifact_create.go [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+	connectcustompluginv1 "github.com/confluentinc/ccloud-sdk-go-v2/connect-custom-plugin/v1"
+
+	pcmd "github.com/confluentinc/cli/v3/pkg/cmd"
+	"github.com/confluentinc/cli/v3/pkg/examples"
+	"github.com/confluentinc/cli/v3/pkg/output"
+)
+
+type pluginCreateOut struct {
+	Name           string `human:"Name" serialized:"name"`
+	Id             string `human:"Plugin ID" serialized:"id"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:19]:
```suggestion
	Id             string `human:"ID" serialized:"id"`
```

File: internal/flink/command_artifact_create.go [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+	ConnectorClass string `human:"Version ID" serialized:"connector_class"`
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:20]: Please make human and serialized labels match

File: internal/flink/command_artifact_create.go [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+		RunE:  c.createArtifact,
+		Example: examples.BuildExampleString(
+			examples.Example{
+				Text: `Create Flink UDF artifact "my-flink-artifact".`,
+				Code: "confluent flink artifact create my-flink-artifact --artifact-file /Users/xyz/Documents/config/plugin.jar",
+			},
+		),
+	}
+
+	cmd.Flags().String("artifact-file", "", "JAR Flink UDF artifact file.")
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:39]:
```suggestion
	cmd.Flags().String("artifact-file", "", "Flink UDF artifact JAR file.")
```

File: internal/flink/command_artifact_describe.go (hunk 15 above)
Comment (brianstrauch) [reply internal/flink/command_artifact_describe.go:19]: Interesting, can we fix the command name?

File: internal/flink/command_artifact_list.go [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+
+func (c *command) list(cmd *cobra.Command, _ []string) error {
+	plugins, err := c.V2Client.ListCustomPlugins()
+	if err != nil {
+		return err
+	}
+
+	list := output.NewList(cmd)
+	for _, plugin := range plugins {
+		if strings.ToLower(*plugin.ConnectorType) != "flink-udf" {
```
Comment (brianstrauch) [inline internal/flink/command_artifact_list.go:46]: Shouldn't the backend return a consistent format? Why do we need to lower-case this?
```suggestion
		if strings.ToLower(plugin.GetConnectorType()) != "flink-udf" {
```

File: internal/flink/utils.go [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/dghubble/sling"
+)
+
+func uploadFile(url, filePath string, formFields map[string]any) error {
+	var buffer bytes.Buffer
+	writer := multipart.NewWriter(&buffer)
+
+	for key, value := range formFields {
+		if strValue, ok := value.(string); ok {
+			_ = writer.WriteField(key, strValue)
```
Comment (brianstrauch) [inline internal/flink/utils.go:21]: Why are we ignoring this error?

File: internal/flink/command_artifact.go (hunk 16 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go]: updated

File: internal/flink/command_artifact.go [hunk 27]
```diff
@@ -19,7 +19,7 @@ type flinkArtifactOut struct {
 func (c *command) newArtifactCommand() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:         "artifact",
-		Short:       "Manage Flink UDF artifact",
+		Short:       "Manage Flink UDF artifact.",
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:22]:
```suggestion
		Short:       "Manage Flink UDF artifacts.",
```

File: internal/flink/command_artifact_create.go [hunk 28]
```diff
@@ -30,14 +30,14 @@ func (c *command) newCreateCommand() *cobra.Command {
 		RunE:  c.createArtifact,
 		Example: examples.BuildExampleString(
 			examples.Example{
-				Text: `Create Flink UDF artifact "my-flink-artifact".`,
+				Text: `Create Flink artifact "my-flink-artifact".`,
 				Code: "confluent flink artifact create my-flink-artifact --artifact-file /Users/xyz/Documents/config/plugin.jar",
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:34]: I think most people would use a relative path here, so can this example default to that?

File: test/fixtures/output/flink/help.golden [hunk 29]
```diff
@@ -4,6 +4,7 @@ Usage:
   confluent flink [command]
 
 Available Commands:
+  artifact     Manage Flink UDF artifact
```
Comment (brianstrauch) [inline test/fixtures/output/flink/help.golden]:
```suggestion
  artifact     Manage Flink UDF artifacts.
```

File: internal/flink/command_artifact.go [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+
+func printTable(cmd *cobra.Command, plugin connectcustompluginv1.ConnectV1CustomConnectorPlugin) error {
+	table := output.NewTable(cmd)
+	// Flink UDF artfact lifecycle management reuses custom connector plugin api for EA customer only
+	// Version ID will be surfaced by `ConnectorClass` for EA
+	// For Flink GA, Flink team will have public documented API for this. Tracking it by internal ticket FRT-334
+	table.Add(&flinkArtifactOut{
+		Name:          plugin.GetDisplayName(),
+		PluginId:      plugin.GetId(),
+		VersionId:     plugin.GetConnectorClass(),
```
Comment (brianstrauch) [inline internal/flink/command_artifact.go:42]: Do we plan to change this field to "VersionId" in the backend?

File: internal/flink/command_artifact_create.go [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+		return err
+	}
+	artifactFile, err := cmd.Flags().GetString("artifact-file")
+	if err != nil {
+		return err
+	}
+
+	extension := strings.ToLower(strings.TrimPrefix(filepath.Ext(artifactFile), "."))
+	if extension != "jar" {
+		return fmt.Errorf(`only file extensions ".jar" is allowed`)
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:63]:
```suggestion
		return fmt.Errorf(`only ".jar" file extension is allowed`)
```

File: internal/flink/command_artifact_create.go [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+	if err := uploadFile(resp.GetUploadUrl(), artifactFile, resp.GetUploadFormData()); err != nil {
+		return err
+	}
+
+	createArtifactRequest := connectcustompluginv1.ConnectV1CustomConnectorPlugin{
+		DisplayName:   connectcustompluginv1.PtrString(displayName),
+		Description:   connectcustompluginv1.PtrString(description),
+		ConnectorType: connectcustompluginv1.PtrString("flink-udf"),
+		UploadSource: &connectcustompluginv1.ConnectV1CustomConnectorPluginUploadSourceOneOf{
+			ConnectV1UploadSourcePresignedUrl: connectcustompluginv1.NewConnectV1UploadSourcePresignedUrl("PRESIGNED_URL_LOCATION", resp.GetUploadId()),
```
Comment (brianstrauch) [inline internal/flink/command_artifact_create.go:84]: Do you mind using an initializer struct here instead of calling this function? (It's a bit more explicit that way- I'm not sure what field "PRESIGNED_URL_LOCATION" is for)

File: internal/flink/command_artifact_list.go [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+
+func (c *command) list(cmd *cobra.Command, _ []string) error {
+	plugins, err := c.V2Client.ListCustomPlugins()
+	if err != nil {
+		return err
+	}
+
+	list := output.NewList(cmd)
+	for _, plugin := range plugins {
+		if strings.ToLower(plugin.GetConnectorType()) != "flink-udf" {
```
Comment (brianstrauch) [inline internal/flink/command_artifact_list.go:46]: Why are we using `strings.ToLower()`? Does the backend not return consistent casing?

File: internal/flink/command_artifact.go (hunk 30 above)
Comment (byang-confluent) [reply internal/flink/command_artifact.go:42]: no, in GA there will be new flink api to handle this tracked by ticket FRT-334

Comment (byang-confluent) [reply internal/flink/command_artifact.go:42]: custom connector api will remain unchanged.

File: internal/flink/command_artifact_create.go (hunk 32 above)
Comment (byang-confluent) [reply internal/flink/command_artifact_create.go:84]: it is part of upload api https://docs.confluent.io/cloud/current/api.html#tag/Presigned-Urls-(connectv1)/operation/presigned-upload-urlConnectV1PresignedUrl

Comment (brianstrauch) [reply internal/flink/command_artifact_create.go:84]: Thanks, can we still define this as a struct so it's more clear for people who look at this code in the future?

Comment (byang-confluent) [reply internal/flink/command_artifact_create.go]: yep. already updated in the latest commit

**Review by brianstrauch — APPROVED** [review]

File: internal/flink/command_artifact_create.go [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+	if err != nil {
+		return err
+	}
+	artifactFile, err := cmd.Flags().GetString("artifact-file")
+	if err != nil {
+		return err
+	}
+
+	extension := strings.ToLower(strings.TrimPrefix(filepath.Ext(artifactFile), "."))
+	if extension != "jar" {
```
[the remark this answers is not in the document]
Comment (brianstrauch) [reply internal/flink/command_artifact_create.go:62]: Let's try to prevent extremely long lines

**Review by brianstrauch — APPROVED** [review]
+1 assuming this PR is available in production (you haven't marked the checkbox in the description)

**Comment by brianstrauch** [comment]
Please add release notes prefixed with "[Early Access]" as well. Also FYI, your GitHub commits are unverified- you should take a look at if you've set up `git` correctly.

**Comment by byang-confluent** [comment]
@brianstrauch yes, all flink command are gated with `cli.flink` LD feature flag. so this PR is OK in production as it's hidden behind the same flag.
